pandas/io/parsers.py

```diff
@@ -2139,6 +2162,52 @@ def _maybe_parse_dates(self, values, index, try_parse_dates=True):
         return values
 
 
+class ArrowParserWrapper(ParserBase):
+    """
+
+    """
+
+    def __init__(self, src, **kwds):
+        self.kwds = kwds
+        self.src = src
+        kwds = kwds.copy()
+
+        ParserBase.__init__(self, kwds)
+
+        # #2442
+        kwds["allow_leading_cols"] = self.index_col is not False
+
+        # GH20529, validate usecol arg before TextReader
+        self.usecols, self.usecols_dtype = _validate_usecols_arg(kwds["usecols"])
+
+    def read(self, nrows=None):
+        pyarrow = import_optional_dependency(
+            "pyarrow.csv", extra="pyarrow is required to use arrow engine"
+        )
+        table = pyarrow.read_csv(
+            self.src,
+            read_options=pyarrow.ReadOptions(
+                skip_rows=self.kwds.get("skiprows"), column_names=self.names
+            ),
+            parse_options=pyarrow.ParseOptions(
+                delimiter=self.kwds.get("delimiter"),
+                quote_char=self.kwds.get("quotechar"),
+            ),
+            convert_options=pyarrow.ConvertOptions(
+                include_columns=self.usecols, column_types=self.kwds.get("dtype")
+            ),
+        )
+        if nrows:
+            table = table[:nrows]
```

Comment:
I would rather raise an error here that `nrows` is not supported, as users will expect this to give less memory overhead, while in this way it is still reading the full file in memory.

Reply:
@jorisvandenbossche : I'm actually ambivalent here.

True as your point may be, we do this already with the Python engine.  It would actually be "consistent" with existing behavior there.

Reply:
> On the other hand, we do not support this type of nrows behavior with the C engine 

Are you sure? The docs don't say anything about it being unsupported by default, and you also don't get a fallback warning

Reply:
@jorisvandenbossche: I'm so sorry.  I completely confused `nrows` with `skipfooter` when I wrote that (tired brain, need more coffee 😁).

I 100% agree with your [comment](https://github.com/pandas-dev/pandas/pull/31817#discussion_r378333329).